### ISSUE TITLE
EOS dev env docs

### ISSUE THREAD

**Issue by PVince81** [owncloud/ocis#361]
Adjust https://owncloud.github.io/ocis/eos/ with steps for anyone to setup EOS locally for debugging.

Using https://github.com/owncloud-docker/compose-playground/tree/master/examples/eos-compose/  or other.

This is crucial for everyone to be able to start debugging and testing against EOS.

- [ ] REQUIRES EOS in CI https://github.com/owncloud/ocis/issues/253 to be able to commit

@felixboehm @butonic @individual-it @micbar

**Comment by PVince81** [comment]
note: even if not fully complete it's ok to add some TODOs in the docs. maybe other people can help there as well

**Comment by butonic** [comment]
First step is https://github.com/owncloud/ocis/pull/349 which lays the foundation.

The idea is to use the [eos-ocis-dev](https://github.com/owncloud-docker/eos-stack/tree/master/eos-ocis-dev) container to build the binary. Building a binary that has all necessary LDAP libs in place to look up a user using the os is tricky as it requires cgo. Getting rid of that is tracked in https://github.com/owncloud/ocis-reva/issues/331

Left tasks i see:
- [ ] update debugging.md to reflect how to use the eos-ocis-dev container to run the debug binary
  - for vscode
  - for the intellij go ide
  - for vim
- [ ] reorganize the debugging.md, development.md and eos.md under a common development section.

**Comment by PVince81** [comment]
Docs are in https://owncloud.github.io/ocis/eos/.

Some topics from me:
- [ ] clarify that one needs a local checkout of "ocis", "reva", etc with go.mod replace edit + docker-compose volumes configured
- [ ] run `make clean` locally to make sure env is clean before running docker-compose to avoid compilation mix-up
- [ ] how to restart/recompile after changes ?

- [ ] part 2 "LDAP support" and part 3: is this optional or mandatory ? maybe add a sentence "if you need a different namespace for the frontend, do this".

**Comment by PVince81** [comment]
fancy idea: add targets to the makefile that runs "docker-compose up" and other commands like rebuilding inside the container :-)

**Comment by PVince81** [comment]
- [ ] cache go modules from deps to accelerate rebuilds (need the go cache folder on a volume?)

**Comment by PVince81** [comment]
@butonic @felixboehm see checkboxes above. Please raise tickets accordingly for the non-trivial tasks that need more time and future scheduling.

for the others I guess it could be done right away

**Comment by PVince81** [comment]
with latest ocis master there is some strange behavior:

- upload doesn't work and fails with a 500 error due to permission denied, see https://github.com/owncloud/product/issues/128#issuecomment-677727082
- when I login as einstein, I see the Webdav requests being done to 4c510ada-c86b-4815-8820-42cdf82c3d51, however in the web UI I see the contents of `docker-compose exec ocis eos ls -l /eos/dockertest/reva/users/e/einstein` despite the existence of the `docker-compose exec ocis eos ls -l /eos/dockertest/reva/users/4/4c510ada-c86b-4815-8820-42cdf82c3d51/` folder. But the latter is empty.

So maybe there's a user / user id mismatch somewhere ??

**Comment by PVince81** [comment]
- [ ] note: to start over, use `docker-compose down -v --rmi all --remove-orphans`

**Comment by PVince81** [comment]
never mind, I just noticed that the docs have changed so I need to change my script as well to use the new layout...

**Comment by PVince81** [comment]
- [ ] document how to access the meta folders like the ones where the chunks are stored.
the `docker-compose exec mgm-master eos` only gives access to the same files like in the web UI but not the meta ones.

**Comment by jnweiger** [comment]
>     * [ ]  add hint that the ocis container will be compiling the go binaries the first time it is run, so need to wait before proceeding to the LDAP step

I have good success with the following wait functions: 
```
wait_for_ocis () {
  ## it compiles code upon first start. this can take ca 6 minutes.
  while true; do
    # expect to see "Starting server ... 0.0.0.0:9200" in the logs.
    docker-compose logs --tail=10 ocis
    if [ -n "\$(docker-compose logs ocis | grep 'Starting server' | grep 0.0.0.0:9200)" ]; then
      break
    fi
    echo " ... waiting for 0.0.0.0:9200 ..."
    sleep 15;
  done
}

wait_for_ldap () {
  # expect nslcd is running.
  while test -z "\$(docker-compose exec ocis ps -ef | grep nslcd | grep -v grep)"; do
    echo "waiting for nslcd ...";
    sleep 3;
  done
}

wait_for_eos_fst () {
  # expect (at least) four fst entries.
  while [ "\$(docker-compose exec ocis eos fs ls -m | grep host=fst | wc -l)" -lt 4 ]; do
    echo "waiting for four fst to appear in 'eos fs ls' ..."
    sleep 3;
    docker-compose exec ocis eos fs ls
  done
}

wait_for_eos_health () {
  echo "Expect to see 'online', 'ok', 'fine', 'default.0' here:"
  for i in 1 2 3 4 5 6 7 8 9 0; do
    # immediately after start, default.0 is shown as 0B free and 'full'
    docker-compose exec ocis eos health -a
    if [ -z "\$(docker-compose exec ocis eos health -m | grep status=full)" ]; then
      break     # nothing is full, carry on.
    fi
    sleep 2
  done
}
```

`wait_for_ocis` should be called before starting ldap. Then call `wait_for_ldap` before using ldap accounts, and so on....

**Comment by jnweiger** [comment]
With ocis-1.0.0rc3 the eos does not work by default. We still need to manually bring it from offline state to online:

```
wait_for_eos_fst
# expect to see stat.active=online four times!
while [ "\$(docker-compose -f $compose_yml exec ocis eos fs ls -m | grep stat.active=online | wc -l)" -lt 4 ]; do
  docker-compose -f $compose_yml exec mgm-master eos space set default on
  sleep 5 
done
wait_for_eos_health

**Comment by phil-davis** [comment]
@dpakach do you know the current working EOS setup and can help get docs up-to-date?

**Comment by jnweiger** [comment]
As discussed today with @butonic, both versions of the docs are outdated. We have identified a list of steps that I should not do, although they were documented.

**Comment by refs** [comment]
superseded by https://github.com/owncloud/ocis/issues/815
